### PR TITLE
Support container recreate in kruise-daemon and add e2e tests

### PR DESCRIPTION
**Pull request by FillZpp**
Signed-off-by: FillZpp <FillZpp.pub@gmail.com>

### Ⅰ. Describe what this PR does
Support container recreate in kruise-daemon and add e2e tests.

### Ⅴ. Special notes for reviews

1. Main logic: `pkg/daemon/containerrecreate/crr_daemon_controller.go`
2. E2E: `test/e2e/apps/containerrecreate.go`
3. Files in `pkg/daemon/kuberuntime` mainly comes from `k8s.io/kubernetes/pkg/kubelet/kuberuntime`, and they all have `Copyright 2016 The Kubernetes Authors.` in header. The `k8s.io/kubernetes/pkg/kubelet/kuberuntime` has too many dependencies, we can not use it directly.

### CODE REVIEW

File: pkg/daemon/containerrecreate/crr_daemon_controller.go [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+			continue
+		}
+
+		kubeContainerStatus := podStatus.FindContainerStatusByName(state.Name)
+		if kubeContainerStatus == nil {
+			break
+		}
+
+		msg := fmt.Sprintf("Stopping container %s by ContainerRecreateRequest %s", state.Name, crr.Name)
+		err := runtimeManager.KillContainer(pod, kubeContainerStatus.ID, state.Name, msg, nil)
```
Comment (furykerry) [inline pkg/daemon/containerrecreate/crr_daemon_controller.go:378]: is it possible the killing action hangs here ?

**Review by furykerry — CHANGES_REQUESTED** [review]

File: pkg/daemon/containerrecreate/crr_daemon_util.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+				syncContainerStatus.ContainerID == kubeContainerStatus.ID.String() &&
+				syncContainerStatus.Ready {
+				status.Phase = appsv1alpha1.ContainerRecreateRequestSucceeded
+			}
+			statuses = append(statuses, status)
+			continue
+		}
+
+		// for no-running state, we consider it will be recreated soon
+		if kubeContainerStatus.State != kubeletcontainer.ContainerStateRunning {
```
[the remark this answers is not in the document]
Comment (FillZpp) [reply pkg/daemon/containerrecreate/crr_daemon_util.go]: done

File: pkg/daemon/containerrecreate/crr_daemon_util.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+		// already restarted/recreated
+		if kubeContainerStatus.ID.String() != c.StatusContext.ContainerID || kubeContainerStatus.RestartCount > int(c.StatusContext.RestartCount) {
+			status := appsv1alpha1.ContainerRecreateRequestContainerRecreateState{
+				Name:  c.Name,
+				Phase: appsv1alpha1.ContainerRecreateRequestRecreating,
+			}
+			if syncContainerStatus != nil &&
+				syncContainerStatus.ContainerID == kubeContainerStatus.ID.String() &&
+				syncContainerStatus.Ready {
```
[the remark this answers is not in the document]
Comment (FillZpp) [reply pkg/daemon/containerrecreate/crr_daemon_util.go]: we consider the container have been recreated successfully only if the new container has started and ready

File: pkg/daemon/containerrecreate/crr_daemon_controller.go [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+				notSatisfied = true
+				continue
+			}
+			klog.Warningf("Wait for CRR %s/%s resourceVersion expectation over %v", crr.Namespace, crr.Name, duration)
+			resourceVersionExpectation.Delete(crr)
+		}
+
+		switch crr.Status.Phase {
+		case "":
+			if err := c.updateCRRPhase(crr, appsv1alpha1.ContainerRecreateRequestPending); err != nil {
```
[the remark this answers is not in the document]
Comment (FillZpp) [reply pkg/daemon/containerrecreate/crr_daemon_controller.go:303]: done, the picked one will be skipped to update to Pending

File: pkg/daemon/containerrecreate/crr_daemon_controller.go (hunk 1 above)
Comment (FillZpp) [reply pkg/daemon/containerrecreate/crr_daemon_controller.go:361]: the `TerminationGracePeriodSeconds` in mock pod will be the killing deadline

**Review by furykerry — APPROVED** [review]

**Pull request merged**